### PR TITLE
Change password generator to use ColoredPassword

### PR DESCRIPTION
**Pull request by ShirokaiLon**
Passwords displayed in the password generator are not colour coded like they are on other platforms. This pull request changes the password generator password to be colour coded.

https://community.bitwarden.com/t/color-coded-password-not-working-on-android/9594

### CODE REVIEW

**Comment by kspearrin** [comment]
This was intentionally not implemented on this page because it causes performance issues. I don't have a solution at this time.

**Comment by ShirokaiLon** [comment]
I see the issue and belive I've found a solution. Using a HTML formatted string seems to have significantly better performance than FormattedString.

I see no noticable performance impact now.

File: src/App/Utilities/PasswordFormatter.cs [hunk 1]
```diff
@@ -21,17 +21,15 @@ private enum CharType
...
             {
-                return new FormattedString();
+                return "";
             }
 
-            var result = new FormattedString();
-            // Start off with an empty span to prevent possible NPEs. Due to the way the state machine
-            // works, this will actually always be replaced by a new span anyway.
-            var currentSpan = new Span();
+            var result = "";
```
Comment (kspearrin) [inline src/App/Utilities/PasswordFormatter.cs]: use `string.Empty`

File: src/App/Utilities/PasswordFormatter.cs [hunk 2]
```diff
@@ -21,17 +21,15 @@ private enum CharType
             Special
         }
 
-        public static FormattedString FormatPassword(string password)
+        public static string FormatPassword(string password)
         {
             if(password == null)
             {
-                return new FormattedString();
+                return "";
```
Comment (kspearrin) [inline src/App/Utilities/PasswordFormatter.cs]: use `string.Empty`

File: src/App/Utilities/PasswordFormatter.cs [hunk 3]
```diff
@@ -56,24 +54,39 @@ public static FormattedString FormatPassword(string password)
...
                     // default value.
                     switch(currentType)
                     {
+                        // Apply color style to span.
+                        // First two digits of returned hex code contains the alpha,
+                        // which is not supported in HTML color,
+                        // so we need to cut those out.
                         case CharType.Number:
-                            currentSpan.TextColor = (Color)Application.Current.Resources["PasswordNumberColor"];
+                            result += $"<span style=\"color:#{((Color)Application.Current.Resources["PasswordNumberColor"]).ToHex().Substring(2)}\">";
```
Comment (kspearrin) [inline src/App/Utilities/PasswordFormatter.cs]: Let's capture the hex codes as strings in variables outside of the for loop and then just use them here.

File: src/App/Utilities/PasswordFormatter.cs [hunk 4]
```diff
@@ -56,24 +54,39 @@ public static FormattedString FormatPassword(string password)
...
                             break;
                         case CharType.Special:
-                            currentSpan.TextColor = (Color)Application.Current.Resources["PasswordSpecialColor"];
+                            result += $"<span style=\"color:#{((Color)Application.Current.Resources["PasswordSpecialColor"]).ToHex().Substring(2)}\">";
```
Comment (kspearrin) [inline src/App/Utilities/PasswordFormatter.cs]: Let's capture the hex codes as strings in variables outside of the for loop and then just use them here.

File: src/App/Utilities/PasswordFormatter.cs [hunk 5]
```diff
@@ -56,24 +54,39 @@ public static FormattedString FormatPassword(string password)
                 // If the char type changed, build a new span to append the text to.
                 if(charType != currentType)
                 {
-                    currentSpan = new Span();
-                    result.Spans.Add(currentSpan);
+                    // Close off previous span.
+                    if (currentType != CharType.None)
```
Comment (kspearrin) [inline src/App/Utilities/PasswordFormatter.cs]: Shouldn't this be `currentType != CharType.None || currentType != CharType.Normal`?

File: src/App/Utilities/PasswordFormatter.cs [hunk 6]
```diff
@@ -56,24 +54,39 @@ public static FormattedString FormatPassword(string password)
...
+                            result += $"<span style=\"color:#{((Color)Application.Current.Resources["PasswordSpecialColor"]).ToHex().Substring(2)}\">";
                             break;
                     }
                 }
-                currentSpan.Text += c;
+                result += c;
             }
+
+            // Close off last span.
+            if (currentType != CharType.Normal)
```
Comment (kspearrin) [inline src/App/Utilities/PasswordFormatter.cs]: Shouldn't this be the same as above? `currentType != CharType.None || currentType != CharType.Normal`?

**Comment by kspearrin** [comment]
Awesome! We can also apply this to the passwords on the password history and password generator history page.

**Comment by ShirokaiLon** [comment]
Good suggestions :+1:

I'll have those things fixed, and will apply it to the history pages tomorrow.

**Comment by ShirokaiLon** [comment]
Used a converter class for the history pages. Looks clean to me, but not sure if there's a better way.

File: src/App/Utilities/PasswordFormatter.cs (hunk 5 above)
Comment (ShirokaiLon) [reply src/App/Utilities/PasswordFormatter.cs]: Indeed. Except && instead of ||

File: src/App/Utilities/PasswordFormatter.cs (hunk 6 above)
Comment (ShirokaiLon) [reply src/App/Utilities/PasswordFormatter.cs]: Indeed. Except && instead of ||

**Pull request merged**